### PR TITLE
 RavenDB-13321: fix concurrency in test

### PR DESCRIPTION
**Pull request by garayx**

### CODE REVIEW

File: src/Raven.Server/Documents/DatabasesLandlord.cs [hunk 1]
```diff
@@ -428,6 +460,12 @@ public bool IsDatabaseLoaded(StringSegment databaseName)
             return false;
         }
 
+        public bool DatabaseRecordExists(string databaseName)
+        {
+            var record = _serverStore.LoadDatabaseRecord(databaseName, out _);
```
Comment (ppekrol) [inline src/Raven.Server/Documents/DatabasesLandlord.cs]: better to use `ReadRawDatabase` from ServerStore.Cluster

File: src/Raven.Client/Documents/Identity/AsyncHiLoIdGenerator.cs [hunk 2]
```diff
@@ -76,7 +77,18 @@ public RangeValue(long min, long max)
         /// <returns></returns>
         public Task<string> GenerateDocumentIdAsync(object entity)
         {
-            return NextIdAsync().ContinueWith(task => GetDocumentIdFromId(task.Result));
+            return NextIdAsync().ContinueWith(task =>
```
Comment (ppekrol) [inline src/Raven.Client/Documents/Identity/AsyncHiLoIdGenerator.cs]: What is the relation between this failing test and this fix?
Would it be better to do await NextIdAsync and then execute second operation?

File: src/Raven.Server/Documents/DatabasesLandlord.cs [hunk 3]
```diff
@@ -168,7 +168,32 @@ private void HandleClusterDatabaseChanged(string databaseName, long index, strin
             }
         }
 
-        private void UnloadDatabase(string databaseName)
+        private void UnloadDatabase(string databaseName, bool dbRecordIsNull = false)
+        {
+            if (dbRecordIsNull)
+            {
+                UnloadDatabaseInternal(databaseName);
```
Comment (ppekrol) [inline src/Raven.Server/Documents/DatabasesLandlord.cs:176]: Need a description here explaining the situation

File: src/Raven.Server/Documents/DatabasesLandlord.cs [hunk 4]
```diff
@@ -178,6 +203,13 @@ private void UnloadDatabase(string databaseName)
             {
                 // this is already in the process of being deleted, we can just exit and let the other thread handle it
             }
+            catch (Exception e)
+            {
```
Comment (ppekrol) [inline src/Raven.Server/Documents/DatabasesLandlord.cs:236]: We should rethrow here? Now we are ignoring all the exceptions.

File: src/Raven.Client/Documents/Identity/AsyncHiLoIdGenerator.cs [hunk 5]
```diff
@@ -76,7 +77,18 @@ public RangeValue(long min, long max)
...
+            {
+                try
+                {
+                    return GetDocumentIdFromId(task.Result);
+                }
+                catch (Exception e)
```
Comment (ayende) [inline src/Raven.Client/Documents/Identity/AsyncHiLoIdGenerator.cs]: This obfuscate the actual exception stack trace. 
I don't follow why this is needed?

File: src/Raven.Server/Documents/DatabasesLandlord.cs [hunk 6]
```diff
@@ -168,7 +168,32 @@ private void HandleClusterDatabaseChanged(string databaseName, long index, strin
...
+            if (DatabasesCache.TryGetValue(databaseName, out var databaseTask) == false)
+            {
+                // database was already unloaded or deleted
+                return;
+            }
+
+            if (databaseTask.IsCompleted == false)
+            {
+                // This case is when an unload was issued prior to the actual loading of a database.
+                databaseTask.ContinueWith(t => { UnloadDatabaseInternal(databaseName); });
```
Comment (ayende) [inline src/Raven.Server/Documents/DatabasesLandlord.cs]: Here you need to check that the load itself was _successful_.
Another issue here is that here you are handling the unload in an _async_ manner.
I'm not sure that this is the right thing to do, and it may cause us concurrency issues.
In some cases, we'll issue a request to unload, but it takes time for this to be processed, but the request that was supposed to unload already returned, leading to file busy issues?

File: src/Raven.Server/Documents/DatabasesLandlord.cs (hunk 4 above)
Comment (ayende) [reply src/Raven.Server/Documents/DatabasesLandlord.cs:236]: Yes, I don't like this. The error handling above is very explicit about why it is handling this. Why add this `catch`?

File: src/Raven.Server/Web/System/AdminDatabasesHandler.cs [hunk 7]
```diff
@@ -260,7 +260,8 @@ public async Task Put()
                     throw licenseLimit;
                 }
 
-                if (ServerStore.DatabasesLandlord.IsDatabaseLoaded(name) == false)
+                if (ServerStore.DatabasesLandlord.IsDatabaseLoaded(name) == false && 
+                    ServerStore.DatabasesLandlord.DatabaseRecordExists(name) == false)
```
Comment (ayende) [inline src/Raven.Server/Web/System/AdminDatabasesHandler.cs]: Why did you add the second check? Any update to the actual database record will not trigger this now, would it?

**Review by ayende — CHANGES_REQUESTED** [review]

File: src/Raven.Client/Documents/Identity/AsyncHiLoIdGenerator.cs (hunk 2 above)
Comment (grisha-kotler) [reply src/Raven.Client/Documents/Identity/AsyncHiLoIdGenerator.cs]: This threw `AggregatedException` instead of `DatabaseDisabledException` as expected.
Before the change the `HiLo` request succeeded while the `SaveChanges` with `DatabaseDisabledException`.

File: src/Raven.Server/Web/System/AdminDatabasesHandler.cs (hunk 7 above)
Comment (grisha-kotler) [reply src/Raven.Server/Web/System/AdminDatabasesHandler.cs]: It should be only triggered for newly created databases on this node.

File: src/Raven.Client/Documents/Identity/AsyncHiLoIdGenerator.cs (hunk 2 above)
Comment (ayende) [reply src/Raven.Client/Documents/Identity/AsyncHiLoIdGenerator.cs]: So handle that at the caller, or turn this into an `async` method, which will do that too

File: src/Raven.Server/Web/System/AdminDatabasesHandler.cs (hunk 7 above)
Comment (ayende) [reply src/Raven.Server/Web/System/AdminDatabasesHandler.cs]: Why does that matter?

File: src/Raven.Server/Documents/DatabasesLandlord.cs (hunk 1 above)
Comment (garayx) [reply src/Raven.Server/Documents/DatabasesLandlord.cs]: done

File: src/Raven.Server/Documents/DatabasesLandlord.cs (hunk 4 above)
Comment (garayx) [reply src/Raven.Server/Documents/DatabasesLandlord.cs:236]: done

File: src/Raven.Client/Documents/Identity/AsyncHiLoIdGenerator.cs (hunk 5 above)
Comment (garayx) [reply src/Raven.Client/Documents/Identity/AsyncHiLoIdGenerator.cs]: done

File: src/Raven.Server/Web/System/AdminDatabasesHandler.cs (hunk 7 above)
Comment (garayx) [reply src/Raven.Server/Web/System/AdminDatabasesHandler.cs]: removed

File: src/Raven.Server/Documents/DatabasesLandlord.cs (hunk 3 above)
Comment (garayx) [reply src/Raven.Server/Documents/DatabasesLandlord.cs:176]: there is a comment on line 73, I can also add it here.

File: src/Raven.Server/Documents/DatabasesLandlord.cs (hunk 6 above)
Comment (garayx) [reply src/Raven.Server/Documents/DatabasesLandlord.cs]: > Here you need to check that the load itself was successful.

added

File: src/Raven.Server/Documents/DatabasesLandlord.cs [hunk 8]
```diff
@@ -182,14 +182,18 @@ private void UnloadDatabase(string databaseName, bool dbRecordIsNull = false)
...
             }
-            else
+            else if (databaseTask.IsCompleted == false)
             {
-                UnloadDatabaseInternal(databaseName);
+                // This case is when an unload was issued prior to the actual loading of a database.
+                databaseTask.ContinueWith(t =>
+                {
+                    if (databaseTask.IsCompletedSuccessfully)
+                        UnloadDatabaseInternal(databaseName);
```
Comment (ayende) [inline src/Raven.Server/Documents/DatabasesLandlord.cs]: This can be delayed arbitrarily, no?

Comment (garayx) [reply src/Raven.Server/Documents/DatabasesLandlord.cs]: done

File: src/Raven.Client/Documents/Identity/AsyncMultiDatabaseHiLoIdGenerator.cs [hunk 9]
```diff
@@ -25,11 +25,11 @@ public AsyncMultiDatabaseHiLoIdGenerator(DocumentStore store, DocumentConvention
         }
 
 
-        public Task<string> GenerateDocumentIdAsync(string dbName, object entity)
+        public async Task<string> GenerateDocumentIdAsync(string dbName, object entity)
         {
             var db = dbName ?? Store.Database;
             var generator = _generators.GetOrAdd(db, GenerateAsyncMultiTypeHiLoFunc);
-            return generator.GenerateDocumentIdAsync(entity);
+            return await generator.GenerateDocumentIdAsync(entity).ConfigureAwait(false);
```
Comment (ppekrol) [inline src/Raven.Client/Documents/Identity/AsyncMultiDatabaseHiLoIdGenerator.cs]: redundant `await` ?

File: src/Raven.Server/Documents/DatabasesLandlord.cs [hunk 10]
```diff
@@ -168,15 +168,62 @@ private void HandleClusterDatabaseChanged(string databaseName, long index, strin
...
+                // This case is when an unload was issued prior to the actual loading of a database.
+                databaseTask.ContinueWith(t =>
+                {
+                    if (databaseTask.IsCompletedSuccessfully)
+                    {
+                        using (_serverStore.ContextPool.AllocateOperationContext(out TransactionOperationContext context))
+                        using (context.OpenReadTransaction())
+                        {
+                            var databaseRecord = _serverStore.Cluster.ReadRawDatabase(context, databaseName, out _);
+                            databaseRecord.TryGet(nameof(DatabaseRecord.Disabled), out bool dbDisabled);
```
Comment (ppekrol) [inline src/Raven.Server/Documents/DatabasesLandlord.cs]: Check DatabasesInfoNotificationSender.IsDatabaseDisabled implementation. It takes more factors into account (e.g. DatabaseState).

**Review by ppekrol — CHANGES_REQUESTED** [review]

File: src/Raven.Client/Documents/Identity/AsyncMultiDatabaseHiLoIdGenerator.cs (hunk 9 above)
Comment (ayende) [reply src/Raven.Client/Documents/Identity/AsyncMultiDatabaseHiLoIdGenerator.cs]: Yes, we can just pass it through.

File: src/Raven.Client/Documents/Identity/AsyncMultiTypeHiLoIdGenerator.cs [hunk 11]
```diff
@@ -30,27 +29,35 @@ public AsyncMultiTypeHiLoIdGenerator(DocumentStore store, string dbName, Documen
...
+            }
 
-            lock (_generatorLock)
+            await _generatorLock.WaitAsync().ConfigureAwait(false);
+
+            try
             {
                 if (_idGeneratorsByTag.TryGetValue(tag, out value))
-                    return value.GenerateDocumentIdAsync(entity);
+                    return await value.GenerateDocumentIdAsync(entity).ConfigureAwait(false);
```
Comment (ayende) [inline src/Raven.Client/Documents/Identity/AsyncMultiTypeHiLoIdGenerator.cs]: This is likely a problem. 
You are doing the expensive ( network call wait) _while_ you are holding the lock.
This needs to be moved outside the lock

**Review by ayende — CHANGES_REQUESTED** [review]

File: src/Raven.Server/Documents/DatabasesLandlord.cs (hunk 10 above)
Comment (garayx) [reply src/Raven.Server/Documents/DatabasesLandlord.cs]: done

File: src/Raven.Client/Documents/Identity/AsyncMultiTypeHiLoIdGenerator.cs (hunk 11 above)
Comment (garayx) [reply src/Raven.Client/Documents/Identity/AsyncMultiTypeHiLoIdGenerator.cs]: done

File: src/Raven.Client/Documents/Identity/AsyncMultiDatabaseHiLoIdGenerator.cs (hunk 9 above)
Comment (garayx) [reply src/Raven.Client/Documents/Identity/AsyncMultiDatabaseHiLoIdGenerator.cs]: done

File: src/Raven.Client/Documents/Identity/AsyncMultiTypeHiLoIdGenerator.cs (hunk 11 above)
Comment (grisha-kotler) [reply src/Raven.Client/Documents/Identity/AsyncMultiTypeHiLoIdGenerator.cs]: _idGeneratorsByTag is a ConcurrentDictionary can't we use GetOrAdd?

Comment (ayende) [reply src/Raven.Client/Documents/Identity/AsyncMultiTypeHiLoIdGenerator.cs]: no, the `lock` there is fine

**Review by ayende — APPROVED** [review]

**Review by ppekrol — APPROVED** [review]
